File: lib/herb/location.rb
# frozen_string_literal: true

module Herb
  class Location
    attr_reader :start, :end

    def initialize(start_position, end_position)
      @start = start_position
      @end = end_position
    end

    def self.from(start_line, start_column, end_line, end_column)
      new(
        Position.new(start_line, start_column),
        Position.new(end_line, end_column)
      )
    end

    def self.[](...)
      from(...)
    end

    def to_hash
      {
        start: start,
        end: self.end,
      }
    end

    def to_json(*args)
      to_hash.to_json(*args)
    end

    def tree_inspect
      %((location: #{start.tree_inspect}-#{self.end.tree_inspect}))
    end

    def inspect
      %(#<Herb::Location #{tree_inspect}>)
    end
  end
end

File: javascript/packages/core/src/location.ts
import { Position } from "./position.js"
import type { SerializedPosition } from "./position.js"

export type SerializedLocation = {
  start: SerializedPosition
  end: SerializedPosition
}

export class Location {
  readonly start: Position
  readonly end: Position

  static from(location: SerializedLocation): Location
  static from(line: number, column: number, endLine: number, endColumn: number): Location
  static from(locationOrLine: SerializedLocation | number, column?: number, endLine?: number, endColumn?: number): Location {
    if (typeof locationOrLine === "number") {
      const start = Position.from(locationOrLine, column!)
      const end = Position.from(endLine!, endColumn!)

      return new Location(start, end)
    } else {
      const start = Position.from(locationOrLine.start)
      const end = Position.from(locationOrLine.end)

      return new Location(start, end)
    }
  }

  static get zero() {
    return new Location(Position.zero, Position.zero)
  }

  constructor(start: Position, end: Position) {
    this.start = start
    this.end = end
  }

  toHash(): SerializedLocation {
    return {
      start: this.start.toHash(),
      end: this.end.toHash(),
    }
  }

  toJSON(): SerializedLocation {
    return this.toHash()
  }

  treeInspect(): string {
    return `${this.start.treeInspect()}-${this.end.treeInspect()}`
  }

  treeInspectWithLabel(): string {
    return `(location: ${this.treeInspect()})`
  }

  inspect(): string {
    return `#<Herb::Location ${this.treeInspect()}>`
  }

  toString(): string {
    return this.inspect()
  }
}
